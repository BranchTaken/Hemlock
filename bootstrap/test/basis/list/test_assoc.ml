open! Basis.Rudiments
open! Basis
open List
open Format

let test () =
  let assocs = [
    [];
    [(0, 10)];
    [(0, 10); (1, 11)];

    [(0, 10); (0, 11); (1, 12)];
    [(0, 10); (1, 11); (0, 12)];
    [(1, 10); (0, 11); (0, 12)];
    [(0, 10); (1, 11); (1, 12); (2, 13)];
  ] in
  let pp_assoc ppf (a, b) = fprintf ppf "(%a,@ %a)" Uns.pp a Uns.pp b in
  let missing = 3 in
  let cmp = Uns.cmp in
  printf "@[<h>";
  iter assocs ~f:(fun assoc ->
    printf "%a\n"
      (pp pp_assoc) assoc
    ;
    iter assoc ~f:(fun (k, _) ->
      printf "find_hlt/mem %a -> %a / %b\n"
        Uns.pp k
        Uns.pp (Assoc.find_hlt k ~cmp assoc)
        (Assoc.mem k ~cmp assoc)
    );

    printf "find/mem %a -> " Uns.pp missing;
    (match (Assoc.find missing ~cmp assoc),
          (Assoc.mem missing ~cmp assoc); with
      | None, b -> printf "None / %b" b
      | Some v, b -> printf "%a / %b" Uns.pp v b
    );
    printf "\n";

    iter assoc ~f:(fun (k, _) ->
      printf "remove_hlt %a -> %a\n"
        Uns.pp k
        (pp pp_assoc) (Assoc.remove_hlt k ~cmp assoc)
    );
    printf "remove %a -> %a\n"
      Uns.pp missing
      (pp pp_assoc) (Assoc.remove missing ~cmp assoc)
    ;
    printf "map -> %a\n"
      (pp pp_assoc) (Assoc.map assoc ~f:(fun v -> v * 2))
    ;

    printf "inverse -> %a\n"
      (pp pp_assoc) (Assoc.inverse assoc);

    printf "\n"
  );
  printf "@]"

let _ = test ()
