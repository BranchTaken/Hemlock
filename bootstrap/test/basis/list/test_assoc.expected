[]
find/mem 3 -> None / false
remove 3 -> []
map -> []
inverse -> []

[(0, 10)]
find_hlt/mem 0 -> 10 / true
find/mem 3 -> None / false
remove_hlt 0 -> []
remove 3 -> [(0, 10)]
map -> [(0, 20)]
inverse -> [(10, 0)]

[(0, 10); (1, 11)]
find_hlt/mem 0 -> 10 / true
find_hlt/mem 1 -> 11 / true
find/mem 3 -> None / false
remove_hlt 0 -> [(1, 11)]
remove_hlt 1 -> [(0, 10)]
remove 3 -> [(0, 10); (1, 11)]
map -> [(0, 20); (1, 22)]
inverse -> [(11, 1); (10, 0)]

[(0, 10); (0, 11); (1, 12)]
find_hlt/mem 0 -> 10 / true
find_hlt/mem 0 -> 10 / true
find_hlt/mem 1 -> 12 / true
find/mem 3 -> None / false
remove_hlt 0 -> [(0, 11); (1, 12)]
remove_hlt 0 -> [(0, 11); (1, 12)]
remove_hlt 1 -> [(0, 10); (0, 11)]
remove 3 -> [(0, 10); (0, 11); (1, 12)]
map -> [(0, 20); (0, 22); (1, 24)]
inverse -> [(12, 1); (11, 0); (10, 0)]

[(0, 10); (1, 11); (0, 12)]
find_hlt/mem 0 -> 10 / true
find_hlt/mem 1 -> 11 / true
find_hlt/mem 0 -> 10 / true
find/mem 3 -> None / false
remove_hlt 0 -> [(1, 11); (0, 12)]
remove_hlt 1 -> [(0, 10); (0, 12)]
remove_hlt 0 -> [(1, 11); (0, 12)]
remove 3 -> [(0, 10); (1, 11); (0, 12)]
map -> [(0, 20); (1, 22); (0, 24)]
inverse -> [(12, 0); (11, 1); (10, 0)]

[(1, 10); (0, 11); (0, 12)]
find_hlt/mem 1 -> 10 / true
find_hlt/mem 0 -> 11 / true
find_hlt/mem 0 -> 11 / true
find/mem 3 -> None / false
remove_hlt 1 -> [(0, 11); (0, 12)]
remove_hlt 0 -> [(1, 10); (0, 12)]
remove_hlt 0 -> [(1, 10); (0, 12)]
remove 3 -> [(1, 10); (0, 11); (0, 12)]
map -> [(1, 20); (0, 22); (0, 24)]
inverse -> [(12, 0); (11, 0); (10, 1)]

[(0, 10); (1, 11); (1, 12); (2, 13)]
find_hlt/mem 0 -> 10 / true
find_hlt/mem 1 -> 11 / true
find_hlt/mem 1 -> 11 / true
find_hlt/mem 2 -> 13 / true
find/mem 3 -> None / false
remove_hlt 0 -> [(1, 11); (1, 12); (2, 13)]
remove_hlt 1 -> [(0, 10); (1, 12); (2, 13)]
remove_hlt 1 -> [(0, 10); (1, 12); (2, 13)]
remove_hlt 2 -> [(0, 10); (1, 11); (1, 12)]
remove 3 -> [(0, 10); (1, 11); (1, 12); (2, 13)]
map -> [(0, 20); (1, 22); (1, 24); (2, 26)]
inverse -> [(13, 2); (12, 1); (11, 1); (10, 0)]

